### PR TITLE
Feature Request: Add update Advertisement feature

### LINKED ISSUE #1478: Feature Request: Add update Advertisement feature
**Is your feature request related to a problem? Please describe.**
Add update advertisement feature in talawa-api to enable editing and updating advertisement once created

**Describe the solution you'd like**
Add necessary inputs and resolvers

**Describe alternatives you've considered**
NA

**Additional context**
NA

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by Vaishali054**
**What kind of change does this PR introduce?**

Added resolvers, mutations and inputs for updating advertisement

**Issue Number:**

Fixes #1478 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

Video of working feature

https://github.com/PalisadoesFoundation/talawa-api/assets/100770875/2f683b09-170d-4c44-9106-1e7b1e2b7aad

Snapshot of mutation result

Screenshot (91)

**If relevant, did you update the documentation?**

NA

**Does this PR introduce a breaking change?**

No

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -544,6 +544,7 @@ type Mutation {
   unlikeComment(id: ID!): Comment
   unlikePost(id: ID!): Post
   unregisterForEventByUser(id: ID!): Event!
+  updateAdvertisement(data: UpdateAdvertisementData!, id: ID!): Advertisement
```
Comment (xoldyckk) [inline schema.graphql:547]: consolidate the data requirements of a mutation within single variable named `input`, take a look at `UpdateUserTagInput`

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by Vaishali054** [comment]
@palisadoes,

This specific test case passes successfully on my local environment. I haven't made any changes directly related to getAdvertisement. All my commits are associated with updating the advertisement.

Screenshot 2023-12-27 122253

Also on the workflow it says expected { _id: 658bc4c9f5c3ed0ab1c2b1dc, …(6) } to have property "name" with value 'myad', but when I was debugging I encountered that it has the property "name" with required value.
Screenshot 2023-12-27 122137

File: schema.graphql (hunk 1 above)
Comment (Vaishali054) [reply schema.graphql]: @xoldyckk updated the code with the changes

**Comment by palisadoes** [comment]
It may be because our tests are expecting node v14
- https://github.com/PalisadoesFoundation/talawa-api/blob/develop/.github/workflows/pull-request.yml

We have a pending PR to upgrade to v20 so your tests may have to be revisited if this is the cause.

**Comment by Vaishali054** [comment]
Then should I wait for it to get resolved or is there something I can do on my end to fix it?

**Comment by palisadoes** [comment]
Resolve it now (if node is the issue). The person updating to v20 will need to resolve the possibility of failing tests because of the upgrade

File: schema.graphql [hunk 2]
```diff
@@ -913,6 +914,15 @@ type UnauthorizedError implements Error {
   message: String!
 }
 
+input UpdateAdvertisementData {
```
Comment (xoldyckk) [inline schema.graphql:917]: name this as `UpdateAdvertisementInput`

Comment (Vaishali054) [reply schema.graphql]: Made the required changes

**Comment by palisadoes** [comment]
Please merge with the latest upstream. We have merged a few PRs recently

**Comment by palisadoes** [comment]
Please fix the failing tests. Sometimes the codecov check fails erroneously. Verify that the code coverage doesn't fail in the Testing application test

**Comment by Vaishali054** [comment]
@palisadoes, none of the files I've modified are listed in the code coverage report above. The code coverage for the 'Testing Application Test' is 97.7%

Screenshot 2023-12-27 234310

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    if (currentUser === null) {
+      throw new errors.NotFoundError(
+        requestContext.translate(USER_NOT_FOUND_ERROR.MESSAGE),
+        USER_NOT_FOUND_ERROR.CODE,
+        USER_NOT_FOUND_ERROR.PARAM
+      );
+    }
+
+    const add = await Advertisement.findOne({
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:23]: use proper variable names, here it should be `advertisement` for example

File: tests/resolvers/Mutation/updateAdvertisement.spec.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    endDate: "2023-01-31",
+    // Add other fields based on your Advertisement model
+  };
+
+  // Create test advertisement in the database (assuming Advertisement model has create method)
+  const createdAdvertisement = await Advertisement.create(
+    testAdvertisementData
+  );
+
+  return {
```
Comment (xoldyckk) [inline tests/resolvers/Mutation/updateAdvertisement.spec.ts:46]: u can simply return `createdAdvertisement.object()`, also this dummy object generation should probaby be moved here https://github.com/PalisadoesFoundation/talawa-api/blob/main/tests/helpers

look at the files in that folder for examples

Comment (Vaishali054) [reply tests/resolvers/Mutation/updateAdvertisement.spec.ts]: made a new file named advertisement in the helpers' folder for dummy object creation and used proper variables in the updated codebase.

File: src/typeDefs/inputs.ts [hunk 5]
```diff
@@ -282,6 +282,15 @@ export const inputs = gql`
     limit: PositiveInt!
   }
 
+  input UpdateAdvertisementInput {
+    id: ID!
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:286]: make this _id

File: src/typeDefs/inputs.ts [hunk 6]
```diff
@@ -282,6 +282,15 @@ export const inputs = gql`
...
+    name: String
+    link: String
+    type: String
+    startDate: String
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:290]: Didn't notice this, make use of https://the-guild.dev/graphql/scalars/docs/scalars/date for dates, if you need the time information as well make use of https://the-guild.dev/graphql/scalars/docs/scalars/date-time

**Review by xoldyckk — COMMENTED** [review]
@palisadoes These PRs and the issues don't provide enough information about the problem they aim to solve. 

Not specifying the user specific permissions, who is allowed to execute this mutation? The superadmin, admin or a normal user?

The advertisement doesn't seem to have any relation to a user, which is usually accomplished by using a foreign key like `creatorId`.

These are a few examples, business requirements for new or existing features should be discussed and clarified before trying to implement them.

**Comment by Vaishali054** [comment]
While coding the update function, I referenced the create advertisement process, inadvertently leading to a lack of user specification checks when updating. @xoldyckk could you please clarify the user specifications on this. I will add them accordingly. Actually, this issue is an extension of issue 1157 of talawa-admin. User specifications was not mentioned there too.

While working on the advertisement feature, I discovered several bugs in the 'create advertisement' mutation:

1.  It does not verify authentication headers, allowing anyone with API access to create an advertisement without authenticated 
     credentials.
2.  The mutation does not verify the orgId, accepting any input as orgId. This might be assumed to always be valid because it's 
    extracted from the frontend URL.
3.  When creating an advertisement, it successfully proceeds even if the startDate is later than the endDate.
4.   There are user specifications to consider, as mentioned by @xoldyckk.

@palisadoes if you find these valid, I could update these changes in create advertisement along with update advertisement.


Screenshot 2023-12-30 130035

Screenshot 2023-12-30 130104

**Comment by palisadoes** [comment]
1. Good observations. Please go ahead.
1. Do you think it's valid to create an issue to go through all the missions and check for these deficiencies? If so, open a separate issue for that and let others assist in getting the code more secure

**Comment by Vaishali054** [comment]
@palisadoes, I plan to open a separate issue regarding this matter. This will enable others to contribute too.

File: src/typeDefs/inputs.ts (hunk 5 above)
Comment (sumitra19jha) [reply src/typeDefs/inputs.ts:286]: Please do the change as per the comment by @xoldyckk

File: src/typeDefs/inputs.ts (hunk 6 above)
Comment (sumitra19jha) [reply src/typeDefs/inputs.ts:290]: Add a custom scalar called Start Date and make sure that scalar should be greater than Date.now()
If not then throw appropriate error. Here is the doc for help: https://www.apollographql.com/docs/apollo-server/schema/custom-scalars/

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        ADVERTISEMENT_NOT_FOUND_ERROR.PARAM
+      );
+    }
+
+    return await Advertisement.findOneAndUpdate(
+      {
+        _id: args.input.id,
+      },
+      {
+        ...(args.input as any),
```
Comment (sumitra19jha) [inline src/resolvers/Mutation/updateAdvertisement.ts:40]: Please make sure if nothing is provided you throw the error.

**Review by sumitra19jha — COMMENTED** [review]
In terms of specification, lets keep auth only now. Letter when specification is precisely provided we can create an issue for change.

**Comment by Vaishali054** [comment]
@sumitra19jha updated the codebase with all the changes

File: src/typeDefs/scalars.ts [hunk 8]
```diff
@@ -11,9 +12,35 @@ export const scalars = gql`
...
   scalar PhoneNumber
   scalar PositiveInt
+  scalar StartDate
   scalar Time
   scalar URL
   scalar Upload
   scalar JSON
   scalar Any
 `;
+export const StartDateResolver = new GraphQLScalarType({
```
Comment (xoldyckk) [inline src/typeDefs/scalars.ts:22]: Remove the new scalar and do comparisions on `startDate` and `endDate` within the resolver. Scalars are meant for extending the default set of scalars provided by graphql, not for putting business logic in the schema.

File: src/typeDefs/inputs.ts [hunk 9]
```diff
@@ -282,6 +282,15 @@ export const inputs = gql`
     limit: PositiveInt!
   }
 
+  input UpdateAdvertisementInput {
+    _id: ID!
+    name: String
+    link: String
+    type: String
+    startDate: StartDate
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:290]: remove this

File: tests/helpers/advertisement.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import { Advertisement } from "../../src/models";
+
+type AdvertisementTypes = {
```
Comment (xoldyckk) [inline tests/helpers/advertisement.ts:3]: This seems wrong to me, it should be `type AdvertisementTypes = "POPUP" | "MENU" | "BANNER"`

What is meant by other properties here:- https://github.com/PalisadoesFoundation/talawa-api/blame/61d4122b769c91d70b59bda26477b35ef773eb81/src/models/Advertisement.ts#L6

@palisadoes

Comment (palisadoes) [reply tests/helpers/advertisement.ts:3]: Not sure. @SiddheshKukade What was your intention with `AdvertisementTypes`?

Comment (SiddheshKukade) [reply tests/helpers/advertisement.ts:3]: @xoldyckk  it's just to differentiate between a popup and banner add. Start date and end date are basically when the campaign should be promoted in the app.

Comment (xoldyckk) [reply tests/helpers/advertisement.ts:3]: @SiddheshKukade no i meant what's the point of making it an object? The differentiation between types of advertisement can be done using just a discriminated string union like I suggested right?

Comment (xoldyckk) [reply tests/helpers/advertisement.ts:3]: Just noticed that the typescript definition and mongoose schema don't match:-

https://github.com/PalisadoesFoundation/talawa-api/blame/61d4122b769c91d70b59bda26477b35ef773eb81/src/models/Advertisement.ts#L6

https://github.com/PalisadoesFoundation/talawa-api/blame/61d4122b769c91d70b59bda26477b35ef773eb81/src/models/Advertisement.ts#L61

In typescript, `advertisement.type` is an `object` with a single field named `type`, in mongoose model `advertisement.type` is a `string`. The type definition is incorrect.

Comment (Vaishali054) [reply tests/helpers/advertisement.ts:3]: @xoldyckk Do I need to change type in TestAdvertisementType to string as in mongoose model?

Comment (xoldyckk) [reply tests/helpers/advertisement.ts:3]: @Vaishali054 yes

Comment (Vaishali054) [reply tests/helpers/advertisement.ts]: @xoldyckk made all the changes to the codebase

File: src/models/Advertisement.ts [hunk 11]
```diff
@@ -12,7 +12,7 @@ export interface InterfaceAdvertisement {
   orgId: string;
   name: string;
   link: string;
-  type: AdvertisementTypes;
+  type: string;
```
Comment (xoldyckk) [inline src/models/Advertisement.ts:15]: make this a typescript discriminated union

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import { errors, requestContext } from "../../libraries";
+import {
+  ADVERTISEMENT_NOT_FOUND_ERROR,
+  USER_NOT_FOUND_ERROR,
+  INPUT_NOT_FOUND_ERROR,
+  END_DATE_VALIDATION_ERROR,
+  START_DATE_VALIDATION_ERROR,
+} from "../../constants";
+
+export const updateAdvertisement: MutationResolvers["updateAdvertisement"] =
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:12]: who is allowed to execute this mutation? 
1. superadmin
2. superadmin, admin
3. superadmin, admin, user

please confirm this

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  async (_parent, args, _context) => {
+    const { _id, ...otherFields } = args.input;
+
+    //If there is no input
+    if (Object.keys(otherFields).length === 0) {
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:17]: In graphql you can provide `null` for the fields that are nullable in schema, this check wouldn't check against that.

Also, what fields should the user be allowed to update? What should be done if the user tries to update a field to a value that is not intended? One example would be making the `organizationId` `null`. This is unintended.

Think of all the possible scenarios possible in a real world workflow, and then express logic in the resolver accordingly.

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+      new Date(startDate) <= new Date(new Date().toDateString())
+    ) {
+      throw new errors.InputValidationError(
+        requestContext.translate(START_DATE_VALIDATION_ERROR.MESSAGE),
+        START_DATE_VALIDATION_ERROR.CODE,
+        START_DATE_VALIDATION_ERROR.PARAM
+      );
+    }
+
+    //If endDate is less than startDate
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:63]: fix the comment, `less then or equal to`

File: schema.graphql [hunk 15]
```diff
@@ -544,6 +544,7 @@ type Mutation {
   unlikeComment(id: ID!): Comment
   unlikePost(id: ID!): Post
   unregisterForEventByUser(id: ID!): Event!
+  updateAdvertisement(input: UpdateAdvertisementInput!): Advertisement
```
Comment (xoldyckk) [inline schema.graphql:547]: Make a new `type UpdateAdvertisementPayload` for defining shape of the mutation result. Refer to this https://github.com/PalisadoesFoundation/talawa-api/pull/1585#discussion_r1441534417

File: src/resolvers/Mutation/updateAdvertisement.ts (hunk 12 above)
Comment (Vaishali054) [reply src/resolvers/Mutation/updateAdvertisement.ts:12]: @xoldyckk For now, we'll maintain the authentication level as '@auth'. We'll create a new issue to address this matter. According to our use case, this mutation should be allowed for superadmins, admins, and the user who created the advertisement. However, in our current codebase, we don't keep track of the user who created the advertisement when it's initially created. This absence of tracking makes it challenging to verify whether the user attempting to update the advertisement is the original creator. Is this okay or do we need to do something else?

File: src/typeDefs/mutations.ts [hunk 16]
```diff
@@ -200,6 +200,10 @@ export const mutations = gql`
 
     unregisterForEventByUser(id: ID!): Event! @auth
 
+    updateAdvertisement(
+      input: UpdateAdvertisementInput!
+    ): UpdateAdvertisementPayload! @auth
```
Comment (xoldyckk) [inline src/typeDefs/mutations.ts:205]: return type should be nullable so that error propagation doesn't nullify other query results
remove the exclamation mark `: UpdateAdvertisementPayload`

File: src/typeDefs/types.ts [hunk 17]
```diff
@@ -92,6 +92,15 @@ export const types = gql`
     endDate: Date!
   }
 
+  type UpdateAdvertisementPayload {
+    _id: ID
```
Comment (xoldyckk) [inline src/typeDefs/types.ts:96]: this should be this:-
```graphql
type UpdateAdvertisementPayload {
  advertisement: Advertisement
}
```

File: src/typeDefs/types.ts [hunk 18]
```diff
@@ -92,6 +92,15 @@ export const types = gql`
...
+    name: String!
+    link: String!
+    type: String!
```
Comment (xoldyckk) [inline src/typeDefs/types.ts:99]: this should be an enum too, just like it's an enum in the database model, use a graphql enum named `AdvertisementType` and make use of SCREAMING_SNAKE_CASE(if needed to seperate words) for different instances of the enum

File: src/resolvers/Mutation/updateAdvertisement.ts (hunk 12 above)
Comment (xoldyckk) [reply src/resolvers/Mutation/updateAdvertisement.ts:14]: @palisadoes

File: src/typeDefs/inputs.ts [hunk 19]
```diff
@@ -270,6 +270,20 @@ export const inputs = gql`
...
+  input UpdateAdvertisementInput {
+    _id: ID!
+    name: String
+    link: String
+    type: String
+    startDate: Date
+    endDate: Date
+  }
+
+  input UpdateEventProjectInput {
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:282]: why is this added? remove this

File: src/resolvers/Mutation/updateAdvertisement.ts (hunk 12 above)
Comment (palisadoes) [reply src/resolvers/Mutation/updateAdvertisement.ts:14]: Ads management would be by SuperAdmin and Admin users

Comment (xoldyckk) [reply src/resolvers/Mutation/updateAdvertisement.ts:14]: without a role based authorization logic this PR cannot be approved

**Comment by Vaishali054** [comment]
@xoldyckk I have added role-based auth. Can you please take a look?

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+      "name",
+      "link",
+      "type",
+      "startDate",
+      "endDate",
+    ];
+
+    // Check for unintended null values in permitted fields, if all fields are permitted
+    for (const field of Object.keys(args.input)) {
+      if (!permittedFields.includes(field)) {
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:38]: this check is useless, graphql only accepts fields that are defined in the schema

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+
+    const updatedAdvertisementPayload = {
+      _id: updatedAdvertisement._id.toString(), // Ensure _id is converted to String as per GraphQL schema
+      name: updatedAdvertisement.name,
+      orgId: updatedAdvertisement.orgId.toString(),
+      link: updatedAdvertisement.link,
+      type: updatedAdvertisement.type,
+      startDate: updatedAdvertisement.startDate,
+      endDate: updatedAdvertisement.endDate,
+    };
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:121]: `.lean()` gives back a simple javascript object instead of mongoose document, no need for `.toString()` conversions, u can simply return the advertisement

**Comment by palisadoes** [comment]
@Vaishali054 

1. Please fix the conflicting files.
2. Go through and mark all the requests for changes in this page that you have completed as being resolved. It will make the PR easier to follow
3. Make any other suggested changes and let us know when you are done.

The PR seems more complicated than originally planned. I'd like to get this resolved and merged. The conversation has reached over 50 comments and the PR has been open for three weeks.

File: src/resolvers/Mutation/updateAdvertisement.ts (hunk 21 above)
Comment (Vaishali054) [reply src/resolvers/Mutation/updateAdvertisement.ts:121]: Whenever I do that, it gives this error 
Type '{ advertisement: InterfaceAdvertisement; }' is not assignable to type 'UpdateAdvertisementPayload'.
The types of 'advertisement._id' are incompatible between these types.
Type 'ObjectId' is not assignable to type 'string'.

File: src/typeDefs/directives.ts [hunk 22]
```diff
@@ -4,5 +4,5 @@ import { gql } from "graphql-tag";
 export const directives = gql`
   directive @auth on FIELD_DEFINITION
 
-  directive @role(requires: UserType) on FIELD_DEFINITION
+  directive @role(requires: [UserType!]) on FIELD_DEFINITION
```
[the remark this answers is not in the document]
Comment (Vaishali054) [reply src/typeDefs/directives.ts:7]: I though so too, but then I tried using removeOrganization mutation it seems to be working fine
Screenshot (90)

**Comment by Vaishali054** [comment]
@palisadoes On it

**Comment by Vaishali054** [comment]
@palisadoes I am done now. Can you please take a look now? There is one unresolved conversation. The way of implementation mentioned is throwing the error mentioned there which is why I implemented the return in this way.

**Comment by palisadoes** [comment]
Reviewers. This should be complete. Please take a look

- @SiddheshKukade
- @sumitra19jha
- @JamaicanFriedChicken
- @xoldyckk

**Review by SiddheshKukade — APPROVED** [review]
LGTM

**Comment by SiddheshKukade** [comment]
@Vaishali054 can you please put a little video in description showing it's working in action.

**Comment by Vaishali054** [comment]
@SiddheshKukade Updated it.

**Comment by palisadoes** [comment]
@SiddheshKukade is this acceptable now?

**Comment by SiddheshKukade** [comment]
@Vaishali054 looks fine @palisadoes good to merge.

**Review by palisadoes — APPROVED** [review]
